indexer/etl/etl.go

```diff
@@ -0,0 +1,112 @@
+package etl
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/indexer/node"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	defaultLoopInterval     = 5 * time.Second
+	defaultHeaderBufferSize = 500
+)
+
+type ETL struct {
+	log log.Logger
+
+	headerTraversal *node.HeaderTraversal
+	ethClient       *ethclient.Client
+	contracts       []common.Address
+
+	etlBatches chan ETLBatch
```

Comment:
Huge go noob. Can you eli5 what a channel is specific this use case?

Reply:
yea ofc! So basically `etl.go`'s job is to just retrieve the next set of headers and logs. This channel represents where that information gets dumped to. Also someone has to read from the channel before it can dump another batch. So it also acts as a buffer to wait until the pending batches sent in the channel are consumed -- (nit but this is a property of "unbuffered channels". "buffered channels work slightly differently)

`l1_etl` and `l2_etl` listen for batches in their respective channels and indexes them into the database. The indexing part is separate because on L2 we index everything and on L1 we only index blocks with logs. This file helps DRY the extraction process of new `ETLBatch`s